### ISSUE TITLE
Class in implementation not recognized as the same class as its stub

### ISSUE THREAD

**Issue by jonathandung** [astral-sh/ty#3438]
### Summary

See [microsoft/pylance-release#8007](https://github.com/microsoft/pylance-release/issues/8007). Apparently ty has the same issue.

```python
# _1.py
TYPE_CHECKING = False
if TYPE_CHECKING:
    from _2 import Bar
class Foo:
    def __init__(self, bar: Bar, /): bar.foo = self

# _1.pyi
from _2 import Bar
class Foo:
    def __init__(self, bar: Bar): ...

# _2.py
class Bar:
    __slots__ = ('foo',)

# _2.pyi
from _1 import Foo
class Bar:
    foo: Foo
```

```bash
> uvx ty check
error[invalid-assignment]: Object of type `Self@__init__` is not assignable to attribute `foo` of type `Foo`
 --> _1.py:5:38
  |
3 |     from _2 import Bar
4 | class Foo:
5 |     def __init__(self, bar: Bar, /): bar.foo = self
  |                                      ^^^^^^^
  |

Found 1 diagnostic
```

Unrelated, but I find it an eyesore that there is no period following the 'Found 1 diagnostic' line, as ruff has punctuation.

### Version

ty 0.0.35 (bc12d1cad 2026-05-10)

**Comment by carljm** [comment]
Thanks for the report! This is #1306.
